hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -60,6 +68,206 @@ async def _wait_closed(self):
+        if self._closed:
+            total_size = self._write_buffer.offset() + self._write_buffer.size()
+            total_size_str = str(total_size)
+        else:
+            total_size = None
+            total_size_str = '*'
+
+        if self._broken:
+            # note: this retries
+            resp = self._session.put(self._session_url,
+                                     headers={
+                                         'Content-Length': '0',
+                                         'Content-Range': f'bytes */{total_size_str}'
+                                     },
+                                     raise_for_status=False)
+            if resp.status >= 200 or resp.status < 300:
+                assert self._closed and self._write_buffer.size() == 0
+                self._done = True
+                return
+            if resp.status == 308:
+                range = resp.header.get('Range')
+                if range is not None:
+                    new_offset = self._range_upper(range) + 1
+                else:
+                    new_offset = 0
+                self._write_buffer.advance_offset(new_offset)
+                self._broken = False
+            else:
+                assert resp.status >= 400
+                resp.raise_for_status()
+                assert False
+
+        assert not self._broken
+        self._broken = True
+
+        offset = self._write_buffer.offset()
+        if self._closed:
+            n = self._write_buffer.size()
+        else:
+            assert self._write_buffer.size() >= self._chunk_size
+            n = self._chunk_size
+        if n > 0:
+            range = f'bytes {offset}-{offset + n - 1}/{total_size_str}'
+        else:
+            range = f'bytes */{total_size_str}'
+
+        it: FeedableAsyncIterable[bytes] = FeedableAsyncIterable()
+        async with _TaskManager(
+                self._session.put(self._session_url,
+                                  data=aiohttp.AsyncIterablePayload(it),
+                                  headers={
+                                      'Content-Length': f'{n}',
+                                      'Content-Range': range
+                                  },
+                                  raise_for_status=False,
+                                  retry=False)) as put_task:
+            for chunk in self._write_buffer.chunks(n):
+                async with _TaskManager(it.feed(chunk)) as feed_task:
+                    done, _ = await asyncio.wait([put_task, feed_task], return_when=asyncio.FIRST_COMPLETED)
+                    if feed_task not in done:
+                        msg = 'resumable upload chunk PUT request finished before writing data'
+                        log.warning(msg)
+                        raise TransientError(msg)
+
+            await it.stop()
+
+            resp = await put_task
+            if resp.status >= 200 and resp.status < 300:
+                assert self._closed
```

Comment:
Who switches the flag here to `self._closed = True`? That must be the user calling `close`. I think I see now how this is setup where the >= 200 or < 300 status code can only come after the user calls close. Otherwise, the response will be 308.

Reply:
That's done by WriteableStream: https://github.com/hail-is/hail/blob/5b25084d7d8d5ff9908dc48ced1f3583eabd25d2/hail/python/hailtop/aiotools/stream.py#L45

Either when `closed`  is called (which is sync), wait_closed is called (which is async, and closes and then waits for the close to complete), or the async context manager exit is called (which calls wait_closed).